core/src/main/java/hivemall/sketch/hll/ApproxCountDistinctUDAF.java

```diff
@@ -0,0 +1,253 @@
+            }
+            return outputOI;
+        }
+
+        @Override
+        public HLLBuffer getNewAggregationBuffer() throws HiveException {
+            HLLBuffer buf = new HLLBuffer();
+            if (params != null) {
+                buf.reset(params[0], params[1]);
+            }
+            return buf;
+        }
+
+        @SuppressWarnings("deprecation")
+        @Override
+        public void reset(@Nonnull AggregationBuffer agg) throws HiveException {
+            HLLBuffer buf = (HLLBuffer) agg;
+            if (params != null) {
+                buf.reset(params[0], params[1]);
+            } else {
+                buf.hll = null;
+            }
+        }
+
+        @SuppressWarnings("deprecation")
+        @Override
+        public void iterate(@Nonnull AggregationBuffer agg, @Nonnull Object[] parameters)
+                throws HiveException {
+            if (parameters[0] == null) {
+                return;
+            }
+
+            HLLBuffer buf = (HLLBuffer) agg;
+            Object value =
+                    ObjectInspectorUtils.copyToStandardJavaObject(parameters[0], origInputOI);
+            Preconditions.checkNotNull(buf.hll, HiveException.class);
+            buf.hll.offer(value);
+        }
+
+        @SuppressWarnings("deprecation")
+        @Override
+        @Nullable
+        public byte[] terminatePartial(@Nonnull AggregationBuffer agg) throws HiveException {
+            HLLBuffer buf = (HLLBuffer) agg;
+            if (buf.hll == null) {
+                return null;
+            }
+            try {
+                return buf.hll.getBytes();
+            } catch (IOException e) {
+                throw new HiveException(e);
+            }
+        }
+
+        @SuppressWarnings("deprecation")
+        @Override
+        public void merge(@Nonnull AggregationBuffer agg, @Nullable Object partial)
+                throws HiveException {
+            if (partial == null) {
+                return;
+            }
+
+            byte[] data = mergeInputOI.getPrimitiveJavaObject(partial);
+            final HyperLogLogPlus otherHLL;
+            try {
+                otherHLL = HyperLogLogPlus.Builder.build(data);
+            } catch (IOException e) {
+                throw new HiveException("Failed to build other HLL");
+            }
+
+            final HLLBuffer buf = (HLLBuffer) agg;
+            if (buf.hll == null) {
+                buf.hll = otherHLL;
+            } else {
+                try {
+                    buf.hll.addAll(otherHLL);
+                } catch (CardinalityMergeException e) {
+                    throw new HiveException("Failed to merge HLL");
+                }
+            }
+        }
+
+        @SuppressWarnings("deprecation")
+        @Override
+        public LongWritable terminate(@Nonnull AggregationBuffer agg) throws HiveException {
+            HLLBuffer buf = (HLLBuffer) agg;
+
+            long cardinarity = (buf.hll == null) ? 0L : buf.hll.cardinality();
```

Comment:
@myui just noticed: s/cardinarity/cardinality/ also in document

fix sometime in the future

Reply:
@takuti good catch! You can fix it directly in master. No need to create a ticket for this kind of fix.